### PR TITLE
request history since last known messages

### PR DESCRIPTION
**Pull request by max-weller**
when rejoining a room at startup

### CODE REVIEW

**Comment by max-weller** [comment]
Hi,
this pull request implements the `<history since>` tag from [XEP-0045, 7.2.15](https://xmpp.org/extensions/xep-0045.html#enter-managehistory).
I need this feature so I get all messages I missed in a MUC room while I was offline.
Can you merge it?
Thanks

File: libdino/src/service/muc_manager.vala [hunk 1]
```diff
@@ -28,14 +28,14 @@ public class MucManager : StreamInteractionModule, Object {
         stream_interactor.get_module(MessageProcessor.IDENTITY).pre_message_received.connect(on_pre_message_received);
     }
 
-    public void join(Account account, Jid jid, string? nick, string? password) {
+    public void join(Account account, Jid jid, string? nick, string? password, string? history_since) {
```
Comment (fiaxh) [inline libdino/src/service/muc_manager.vala]: Could you please not make history_since an argument here, but instead determine the value in this function? The way it is now, only history for autojoined MUCs but not for manually joined MUCs would be properly handled.

File: libdino/src/service/muc_manager.vala [hunk 2]
```diff
@@ -229,7 +229,10 @@ public class MucManager : StreamInteractionModule, Object {
             foreach (Xep.Bookmarks.Conference bookmark in conferences) {
                 Jid jid = new Jid(bookmark.jid);
                 if (bookmark.autojoin) {
-                    join(account, jid, bookmark.nick, bookmark.password);
+                    string history_since = null;
+                    Conversation? conversation = stream_interactor.get_module(ConversationManager.IDENTITY).get_conversation(jid, account);
+                    if (conversation != null) history_since = conversation.last_active.to_string();
```
Comment (fiaxh) [inline libdino/src/service/muc_manager.vala]: MessageStorage.get_last_message(conversation).time would be a better suited source for that information.

**Comment by max-weller** [comment]
Those are good ideas, I changed it accordingly.

**Comment by max-weller** [comment]
@fiaxh,
what do you think about the pull request with the changes applied?

**Comment by fiaxh** [comment]
Thanks!

**Pull request merged**